### DISCUSSION TITLE
Cpp compiler support for cleaning up data-flow

### DISCUSSION THREAD

**Discussion by Alberth289346** [CorsixTH/CorsixTH#1819] [Research Room]
I managed to build `animview` officially now with help from #1816, but compiling the program didn't give me a useful answer for #1817 (compiler should point out errors in narrowing).

So I tried a different compiler and added more warning options until I actually got useful messages. Currently I use `-Wall -Wextra -std=c++11 -Wconversion -Wsign-conversion -Wno-unused-parameter` and that gives a lot of messages, although it seems with duplicates as some errors are in common include files.

It appears that #1817 isn't even close to fixing although it's a step in the right direction. I tried fixing the other errors too, but it's overwhelming. Many errors in various parts of the code, understanding it is complicated.

While pondering on it during a walk outside, I realized that one problem is that we throw everything into `uint8_t` and friends, but in reality we have many different kinds of data. Colours, pixels, colour channels, indices into containers, sizes of things, etc.
Being able to say 'this value is a pixel' would help enormously in understanding the code, especially if the compiler doesn't allow mixing different kinds of data without us stating it explicitly.

So I wonder if that is possible, have different forms of bytes that you can freely copy and compute on within its own type so we can express how that data flows through the algorithms, without the compiler allowing those bytes to implicitly mix with another form of bytes. Ideally this would all be just in the type system without affecting the compiler output, but I would think it's also feasible if we could do that just for checking correctness of the code, and then for production use replace all byte forms with a common `uint8_t` type and empty conversions between different kinds of bytes.

**Comment by TheCycoONE** [comment]
Making better use of the type system is a great idea, and it's free at run time, there's no reason to remove it from production use.
Of course everything has to go back to primitives or library specific types at the boundaries.

**Comment by Alberth289346** [comment]
Enums seem to do the trick.

``` cpp
#include <cstdint>
#include <cstdio>

enum bpp8_t : std::uint8_t { low_bpp8=0, high_bpp8=0xff };
enum channel_t : std::uint32_t { low_channel=0, high_channel=0xffffffff };

int main() {
    bpp8_t p = static_cast<bpp8_t>(21);
    bpp8_t q = p;
           q = static_cast<bpp8_t>(p | p);
    bpp8_t r = static_cast<bpp8_t>(p * 2);
    std::uint8_t v = r;

    channel_t c = static_cast<channel_t>(13);
    channel_t d = static_cast<channel_t>(r);

    printf("p=%u, q=%u, r=%u, c=%u\n", p, q, r, c);
    printf("size uint8=%lu, bpp8=%lu, channel=%lu\n",
           sizeof(std::uint8_t), sizeof(bpp8_t), sizeof(channel_t));
    return 0;
}
```

Which results in
``` sh
$ g++ -std=c++11  x.cpp
$ ./a.out              
p=21, q=21, r=42, c=42
size uint8=1, bpp8=1, channel=4
```
Converting to the enumeration from a number requires a cast, any computation requires a cast (even something like `p | p` ie with 2 enum values). Converting back to an integer number just works.

**Reply by TheCycoONE** [reply]
This is not nice. You can make your own types (class) that either explicitly or implicitly convert to primitives and have zero runtime overhead. There are much better examples in The C++ Programming Language, but this is a start: https://stackoverflow.com/questions/17793298/c-class-wrapper-around-fundamental-types

**Comment by Alberth289346** [comment]
Yeah, after some more thought I also realized you can write `operator` overloads for the arithmetic code. I am not sure how much that is needed, as you'd do this typically for stored data in buffers etc which isn't modified much.

Being able to specify type conversions is nice though if you use a class/struct. So a next question is about the size of such a struct/class, and then perhaps some first experiments. Any thoughts on a good piece of code for trying this?
